sycl/source/detail/handler_impl.hpp

```diff
@@ -198,6 +198,12 @@ class handler_impl {
 
   // Allocation ptr to be freed asynchronously.
   void *MFreePtr = nullptr;
+
+  // Store information about the kernel arguments.
+  void *MKernelFuncPtr = nullptr;
+  int MKernelNumArgs = 0;
+  detail::kernel_param_desc_t (*MKernelParamDescGetter)(int) = nullptr;
+  bool MKernelIsESIMD = false;
```

Comment:
I wonder if we can just keep that inside `HostKernel` class.

Reply:
Probably?  My original hope here was that we could get rid of `HostKernel`, but I didn't appreciate that we need it in order to extend the lifetime of the kernel function until `handler::finalize` is called.

I'm going to focus on getting a version of things that works and demonstrates the optimizations, but once I reach that point I'm open to talking about where the variables should go.